### PR TITLE
[DA][testing] RFC: Create AutomationConditionTester class

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This is an initial forray into exposing a testing API for AutomationConditions. It's quite minimal at the moment, but contains the necessary functions for a user to test how either new materializations or the passage of time might impact the decisions made. It's inspired by the testing framework we use internally, but with a bit less magic.

One constraint here was that any return value from this must only use public APIs, meaning no `AssetKeyPartitionKey` or `AssetSubset`, hence the somewhat weird shape of the AutomationConditionTesterResult object, and the fact that we just log the evaluation results rather than return them outright.

This contains a single test at the moment, roughly indicating how a user might interface with this, but just wanted to put this out to assess the vibes on this type of pattern.

### CODE REVIEW

**Comment by OwenKephart** [comment]
* **#22292** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22292?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by schrockn** [comment]
I think this is too complicated for a public API. The primary interface should be a simple function with this as an opt-in advanced API.

**Comment by OwenKephart** [comment]
@schrockn hm so the alternative testing path with a single function would look like:

```python

from dagster import execute_automation_tick, AutomationTickResult

def my_test():

    instance=DagsterInstance.ephemeral()

    result = execute_automation_tick(defs=defs, current_time=..., instance=instance)

    assert result.num_requested == 1

    result = execute_automation_tick(defs=defs, current_time=..., cursor=result.cursor, instance=instance)

    assert result.num_requested == 0

```

This isn't the end of the world but is a bit more verbose because you have to explicitly shuttle the instance and cursor between ticks. 

The main thing I was trying to accomplish with the class was to avoid that, but there's definitely no harm in starting lightweight.

**Comment by OwenKephart** [comment]
@schrockn updated the PR to move the class thing out of the public API and add an evaluate_automation_conditions method.

**Comment by schrockn** [comment]
Yeah I think the function version is simpler and easier to understand. If we want to handle `instance` and `context` management I think we could do something like

```python
with in_testing_context(context=context, instance=instance):
    result = execute_automation_tick(defs=defs, current_time=...)
```

or 

```python
with automation_state(context=context, instance=instance, current_time=...) as state:
    result = state.execute_automation_tick(defs=defs)
```

and it just is a `partial` underneath the hood rather than a stateful builder thing.

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition_tester.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def evaluate_automation_conditions(
+    defs: Definitions,
+    instance: DagsterInstance,
+    asset_selection: AssetSelection = AssetSelection.all(),
+    evaluation_time: Optional[datetime.datetime] = None,
+    cursor: Optional[AssetDaemonCursor] = None,
+    logger: Optional[logging.Logger] = None,
+) -> EvaluateAutomationConditionsResult:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition_tester.py:62]: I think we should figure out a way to allow for passing a list of `AssetsDefinitions`

**Review by schrockn — CHANGES_REQUESTED** [review]

**Comment by OwenKephart** [comment]
@schrockn gotcha -- removed the class-based thing, made `defs` accept a `Sequence[AssetsDefinition]`, and added a docstring

File: python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition_tester.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        asset_graph=asset_graph,
+        asset_keys=asset_selection.resolve(asset_graph),
+        asset_graph_view=asset_graph_view,
+        logger=logging.getLogger("dagster.automation_condition_tester"),
+        cursor=cursor or AssetDaemonCursor.empty(),
+        data_time_resolver=data_time_resolver,
+        respect_materialization_data_versions=False,
+        auto_materialize_run_tags={},
+    )
+    results, requested_asset_partitions = evaluator.evaluate()
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition_tester.py:126]: Definitely should be a goal to have this return a range representation soon. cc: @smackesey

**Review by schrockn — APPROVED** [review]
cool thanks

**Pull request merged**